cpp/src/hdbscan/detail/extract.cuh

```diff
@@ -0,0 +1,266 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <label/classlabels.cuh>
+
+#include <cub/cub.cuh>
+
+#include <raft/cudart_utils.h>
+
+#include "membership.cuh"
+#include "select.cuh"
+#include "stabilities.cuh"
+#include "utils.h"
+
+#include <raft/sparse/op/sort.h>
+#include <raft/sparse/convert/csr.cuh>
+
+#include <cuml/cluster/hdbscan.hpp>
+
+#include <raft/label/classlabels.cuh>
+
+#include <algorithm>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace ML {
+namespace HDBSCAN {
+namespace detail {
+namespace Extract {
+
+/**
+ * Union-rank data structure with path compression for
+ * labeling data points based on their farthest ancestors
+ * under root.
+ *
+ * For correctness, it is important that all children are
+ * visited before their parents.
+ * @tparam value_idx
+ */
+template <typename value_idx>
+class TreeUnionFind {
+ public:
+  TreeUnionFind(value_idx size_) : size(size_), data(size_ * 2, 0) {
+    for (int i = 0; i < size; i++) {
+      data[i * 2] = i;
+    }
+  }
+
+  void perform_union(value_idx x, value_idx y) {
+    value_idx x_root = find(x);
+    value_idx y_root = find(y);
+
+    if (data[x_root * 2 + 1] < data[y_root * 2 + 1])
+      data[x_root * 2] = y_root;
+    else if (data[x_root * 2 + 1] > data[y_root * 2 + 1])
+      data[y_root * 2] = x_root;
+    else {
+      data[y_root * 2] = x_root;
+      data[x_root * 2 + 1] += 1;
+    }
+  }
+
+  value_idx find(value_idx x) {
+    if (data[x * 2] != x) {
+      data[x * 2] = find(data[x * 2]);
+    }
+
+    return data[x * 2];
+  }
+
+  value_idx *get_data() { return data.data(); }
+
+ private:
+  value_idx size;
+  std::vector<value_idx> data;
+};
+
+template <typename value_idx, typename value_t>
+void do_labelling_on_host(
+  const raft::handle_t &handle,
+  Common::CondensedHierarchy<value_idx, value_t> &condensed_tree,
+  std::set<value_idx> &clusters, value_idx n_leaves, bool allow_single_cluster,
+  value_idx *labels, value_t cluster_selection_epsilon) {
+  auto stream = handle.get_stream();
+
+  std::vector<value_idx> children_h(condensed_tree.get_n_edges());
+  std::vector<value_t> lambda_h(condensed_tree.get_n_edges());
+  std::vector<value_idx> parent_h(condensed_tree.get_n_edges());
+
+  raft::update_host(children_h.data(), condensed_tree.get_children(),
+                    condensed_tree.get_n_edges(), stream);
+  raft::update_host(parent_h.data(), condensed_tree.get_parents(),
+                    condensed_tree.get_n_edges(), stream);
+  raft::update_host(lambda_h.data(), condensed_tree.get_lambdas(),
+                    condensed_tree.get_n_edges(), stream);
+
+  CUDA_CHECK(cudaStreamSynchronize(stream));
+
+  value_idx size = *std::max_element(parent_h.begin(), parent_h.end());
```

Comment:
Would it make sense to replace this with a thrust call on the device array, or at this stage the graph is expected to be condensed to a small size?

Reply:
It won't be as small as the cluster tree, so I went ahead and switched it to thrust/device